addons/udes_stock_refactoring/models/stock_move.py

```diff
@@ -0,0 +1,355 @@
+                - 'assign': assigned, partially_available
+                - 'done': done
+
+           Methods doing a refactor are expected to take a single recordset of
+           moves on which they will act, and to return the recordset of
+           equivalent moves after they have been transformed.
+           The output moves may be identical to the input, may contain none
+           of the input moves, or anywhere in between.
+           The output should contain a functionally similar set of moves.
+        """
+        if stage is not None and stage not in U_STOCK_REFACTOR_STAGES.values():
+            raise UserError(_("Unknown stage for move refactor: %s") % stage)
+        moves = self
+
+        if self.env.context.get("disable_move_refactor"):
+            return moves
+
+        refactor_moves = moves.filtered(
+            lambda m: m.picking_type_id and m.state not in ["draft", "cancel"]
+        )
+
+        if stage is not None:
+            refactor_moves = refactor_moves.filtered(
+                lambda m: U_STOCK_REFACTOR_STAGES[m.state] == stage
+            )
+
+        for picking_type, picking_type_moves in refactor_moves.groupby("picking_type_id"):
+            for stage, stage_moves in picking_type_moves.groupby(
+                lambda m: U_STOCK_REFACTOR_STAGES[m.state]
+            ):
+                refactor_class = self._get_refactor_class(picking_type, stage)
+
+                if refactor_class:
+                    _logger.info(
+                        "Refactoring %s at %s using %s: %s",
+                        picking_type.name,
+                        stage,
+                        refactor_class.name(),
+                        stage_moves.ids,
+                    )
+
+                    new_moves = refactor_class.do_refactor(stage_moves)
+                    if new_moves is not None:
+                        moves -= stage_moves
+                        moves |= new_moves
+
+        return moves
+
+    def _action_confirm(self, *args, **kwargs):
+        """
+        Extend _action_confirm to trigger refactor action.
+
+        Odoos move._action_confirm returns all the moves passed in to it, after
+        merging any it can. In places the return value is used to immediately
+        assign, so any created moves should be returned.
+
+        However, in stock.move._split() in Odoo core, _action_confirm is
+        called on a single move and it expects a single move to be returned
+        (will error out if not). Therefore the refactor at confirm should avoid
+        splitting/creating moves. Luckily this is unlikely to be an issue as
+        it makes little sense to split moves at this stage, before they have
+        stock reserved against them.
+        """
+        res = super(StockMove, self)._action_confirm(*args, **kwargs)
+        post_refactor_moves = res.exists()._action_refactor(stage="confirm")
+
+        if post_refactor_moves != res:
```

Comment:
Could we check if MRP is installed?

Reply:
Good idea, modules are records just like everything else in Odoo so should be possible. Will go look into this.

Reply:
Have added a `get_module_installed` method in `udes_common` and now make use of it here so that the error message is only raised if the `mrp` is installed.

Reply:
Note: ```get_module_installed``` was renamed ```is_module_installed``` after this suggestion: https://github.com/unipartdigital/udes-open/pull/307#pullrequestreview-446292141